lib/fluent/supervisor.rb

```diff
@@ -35,23 +37,109 @@
 end
 
 module Fluent
-  class Supervisor
-    def self.get_etc_passwd(user)
-      if user.to_i.to_s == user
-        Etc.getpwuid(user.to_i)
-      else
-        Etc.getpwnam(user)
-      end
+  module ServerModule
+    def before_run
+      log_opts = {suppress_repeated_stacktrace: config[:suppress_repeated_stacktrace]}
+      log = Supervisor::LoggerInitializer.new(config[:log_path], config[:log_level], config[:chuser], config[:chgroup], log_opts)
+      log.init
+      @logger = $log
+
+      socket_manager_path = ServerEngine::SocketManager::Server.generate_path
+      ServerEngine::SocketManager::Server.open(socket_manager_path)
+      ENV['SERVERENGINE_SOCKETMANAGER_PATH'] = socket_manager_path.to_s
     end
+  end
 
-    def self.get_etc_group(group)
-      if group.to_i.to_s == group
-        Etc.getgrgid(group.to_i)
-      else
-        Etc.getgrnam(group)
+  module WorkerModule
+    def spawn(process_manager)
+      main_cmd = config[:main_cmd]
+      # process_manager.spawn(*Array(cmd))
+      @pm = process_manager.spawn(main_cmd)
+    end
+
+    def after_start
+      install_supervisor_signal_handlers
+    end
+
+    def install_supervisor_signal_handlers
+      trap :USR1 do
+        $log.debug "fluentd supervisor process get SIGUSR1"
+        supervisor_sigusr1_handler
+      end unless ServerEngine.windows?
+    end
+
+    def supervisor_sigusr1_handler
+      # @log.reopen!
+      if pid = @pm.pid
+        Process.kill(:USR1, pid)
+        # don't resuce Erro::ESRSH here (invalid status)
       end
     end
+  end
+
+  def self.load_config(path, params = {})
+    config_mtime = File.mtime(path)
+    # reuse previous config if last load time is within 5 seconds and mtime of the config file is not changed
+    if Time.now - Time.at(params['pre_loadtime'].to_i) < 5 and config_mtime == params['pre_config_mtime']
+      return params['pre_conf']
+    end
+
+    config_fname = File.basename(path)
+    config_basedir = File.dirname(path)
+    config_data = File.read(path)
+    inline_config = params['inline_config']
+    if inline_config == '-'
+      config_data << "\n" << STDIN.read
+    elsif inline_config
+      config_data << "\n" << inline_config.gsub("\\n","\n")
+    end
+    fluentd_conf = Fluent::Config.parse(config_data, config_fname, config_basedir, params['use_v1_config'])
+    system_config = SystemConfig.create(fluentd_conf)
+
+    log_level = system_config.log_level || params['log_level']
```

Comment:
I think we can reflect new setting by reloading config file when dynamic reloading.
If it's wrong, please tell me.